### PR TITLE
dummy-ecall-power: Let ecall do anything

### PR DESCRIPTION
**Pull request by naure**
A variant of #369 with all-powerful ecall instead of NOP.

### CODE REVIEW

File: ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs [hunk 1]
```diff
@@ -67,21 +67,23 @@ impl<E: ExtensionField> DummyConfig<E> {
         circuit_builder: &mut CircuitBuilder<E>,
         codes: InsnCodes,
     ) -> Result<Self, ZKVMError> {
-        let (with_rs1, with_rs2, with_rd) = match (codes.format, codes.kind) {
-            // ECALL reads its syscall_id, then do nothing.
-            (_, InsnKind::EANY) => (true, false, false),
+        // ECALL can do everything.
+        let is_ecall = matches!(codes.kind, InsnKind::EANY);
+        let (with_rs1, with_rs2, with_rd) = match codes.format {
+            _ if is_ecall => (true, true, true),
```
Comment (kunxian-xia) [inline ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs:73]: I felt like we should let the caller to decide how many registers they want to access.

Comment (kunxian-xia) [reply ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs:73]: For example, `ecall/commit` only needs to access X5 and X10 (set `with_rs1 = true, with_rs2 = true`). While complex ecall may use X5 / X10 / X11 by setting `with_rs1 = true, with_rs2 = true, with_rd = true`.

Comment (naure) [reply ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs]: I made it configurable so it is easy to write restricted ecall circuit. The default option is everything at once.

**Comment by naure** [comment]
Clearly this is what we want so I include it in #369 .

File: ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs [hunk 2]
```diff
@@ -34,7 +34,37 @@ impl<E: ExtensionField, I: RIVInstruction> Instruction<E> for DummyInstruction<E
...
+            InsnFormat::S => (true, true, false),
+            InsnFormat::B => (true, true, false),
+            InsnFormat::U => (false, false, true),
+            InsnFormat::J => (false, false, true),
+        };
+        let with_mem_write = matches!(codes.category, InsnCategory::Store) || is_ecall;
+        let with_mem_read = matches!(codes.category, InsnCategory::Load);
+        let branching = matches!(codes.category, InsnCategory::Branch)
+            || matches!(codes.kind, InsnKind::JAL | InsnKind::JALR)
+            || is_ecall;
```
Comment (kunxian-xia) [inline ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs:56]: Most of `Ecall` instructions follow the default branching rule `pc' = pc + 4` except `Ecall/Halt`.

Comment (naure) [reply ceno_zkvm/src/instructions/riscv/dummy/dummy_circuit.rs:56]: That’s ok, who can do more, can do less. I do imagine that is how I will do the complex ecalls: jump to handler code.

**Pull request merged**